File: src/main/java/ru/arc/listeners/RespawnListener.java
package ru.arc.listeners;

import java.util.Set;

import org.bukkit.Location;
import org.bukkit.Tag;
import org.bukkit.World;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.block.Action;
import org.bukkit.event.player.PlayerInteractEvent;
import org.bukkit.event.player.PlayerRespawnEvent;
import org.bukkit.scheduler.BukkitRunnable;
import ru.arc.ARC;
import ru.arc.configs.Config;
import ru.arc.configs.ConfigManager;
import ru.arc.hooks.HookRegistry;

import static ru.arc.util.Logging.error;
import static ru.arc.util.Logging.info;

public class RespawnListener implements Listener {

    private static final Config config = ConfigManager.of(ARC.plugin.getDataPath(), "misc.yml");

    @EventHandler
    public void onRespawn(PlayerRespawnEvent event) {
        info("Respawning player {} reason {} location {}", event.getPlayer().getName(), event.getRespawnReason(), event.getRespawnLocation());
        World world = event.getPlayer().getLocation().getWorld();
        info("World {}", world.getName());
        Set<String> interceptWorlds = config.stringSet("respawn-intercept-worlds");
        if (interceptWorlds.contains(world.getName())) {
            info("Intercepting respawn for player {}", event.getPlayer().getName());
            Location respawnLocation = event.getPlayer().getRespawnLocation();
            info("Respawn location {}", respawnLocation);
            if (respawnLocation == null) return;
            event.setRespawnLocation(respawnLocation);
        }
    }

    @EventHandler
    public void onBedUse(PlayerInteractEvent e) {
        Player player = e.getPlayer();
        if (e.getAction() == Action.RIGHT_CLICK_BLOCK) { // check if they're trying to sleep
            if (e.getClickedBlock() != null && Tag.BEDS.isTagged(e.getClickedBlock().getType())) { // check if the block is a bed
                if ((player.getWorld().getTime() < 12541 || player.getWorld().getTime() > 23458) && !player.getWorld().hasStorm()) {
                    e.setCancelled(true);
                    info("Player {} tried to sleep during the day", player.getName());
                }
                final Location oldRespawn = player.getRespawnLocation();
                HookRegistry.huskHomesHook.hasHome(player)
                        .thenAccept(hasHome -> {
                            info("Player {} has home {}", player.getName(), hasHome);
                            try {
                                if (!hasHome) {
                                    HookRegistry.huskHomesHook.createDefaultHome(player, player.getLocation());
                                    player.sendMessage(config.componentDef("rtp-respawn.bed-create-home", "<green>Ваш <gold>/home<green> установлен здесь! <gray>Чтобы изменить его, используйте команду /sethome"));
                                } else {
                                    new BukkitRunnable() {
                                        @Override
                                        public void run() {
                                            info("Setting respawn location for player {} to {}", player.getName(), oldRespawn);
                                            player.setRespawnLocation(oldRespawn, true);
                                        }
                                    }.runTaskLater(ARC.plugin, 3L);
                                }
                            } catch (Exception ex) {
                                error("Error setting respawn location for player {}", player.getName(), ex);
                            }
                        });

            }
        }
    }


}

File: src/main/kotlin/ru/arc/listeners/RespawnListener.kt
package ru.arc.listeners

import org.bukkit.Tag
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.block.Action
import org.bukkit.event.player.PlayerInteractEvent
import org.bukkit.event.player.PlayerRespawnEvent
import ru.arc.ARC
import ru.arc.config.ConfigManager
import ru.arc.core.sync
import ru.arc.core.ticks
import ru.arc.hooks.HookRegistry
import ru.arc.util.Logging.error
import ru.arc.util.Logging.info

class RespawnListener : Listener {

    private val config = ConfigManager.of(ARC.instance.dataPath, "modules/misc.yml")

    @EventHandler
    fun onRespawn(event: PlayerRespawnEvent) {
        info("Respawning player {} reason {} location {}", event.player.name, event.respawnReason, event.respawnLocation)
        val world = event.player.location.world
        info("World {}", world.name)
        val interceptWorlds = config.stringSet("respawn-intercept-worlds")
        if (interceptWorlds.contains(world.name)) {
            info("Intercepting respawn for player {}", event.player.name)
            val respawnLocation = event.player.getRespawnLocation()
            info("Respawn location {}", respawnLocation)
            if (respawnLocation == null) return
            event.respawnLocation = respawnLocation
        }
    }

    @EventHandler
    fun onBedUse(e: PlayerInteractEvent) {
        val player = e.player
        if (e.action == Action.RIGHT_CLICK_BLOCK) {
            val clickedBlock = e.clickedBlock ?: return
            if (!Tag.BEDS.isTagged(clickedBlock.type)) return
            if ((player.world.time < 12541 || player.world.time > 23458) && !player.world.hasStorm()) {
                e.isCancelled = true
                info("Player {} tried to sleep during the day", player.name)
            }
            val oldRespawn = player.getRespawnLocation()
            val hook = HookRegistry.huskHomesHook ?: return
            val bedLocation = player.location.clone()
            hook.hasHome(player).whenComplete { hasHome, failure ->
                sync {
                    if (failure != null) {
                        error("Error checking home for player {}", player.name, failure)
                        return@sync
                    }
                    info("Player {} has home {}", player.name, hasHome)
                    if (!hasHome) {
                        hook.createDefaultHome(player, bedLocation)
                        player.sendMessage(config.component("rtp-respawn.bed-create-home", "<green>Ваш <gold>/home<green> установлен здесь! <gray>Чтобы изменить его, используйте команду /sethome"))
                    } else {
                        delayed(3.ticks) {
                            info("Setting respawn location for player {} to {}", player.name, oldRespawn)
                            player.setRespawnLocation(oldRespawn, true)
                        }
                    }
                }
            }
        }
    }
}
